src/platforms/esp32/components/avm_builtins/storage_nif.c

```diff
@@ -0,0 +1,340 @@
+    if (!strncmp(part_by_name_prefix, source, part_by_name_len)) {
+        mount_config.allocation_unit_size = CONFIG_WL_SECTOR_SIZE;
+
+        mount = enif_alloc_resource(platform->mounted_fs_resource_type, sizeof(struct MountedFS));
+        if (IS_NULL_PTR(mount)) {
+            RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+        }
+        mount->base_path = strdup(target);
+        mount->mount_type = FATSPIFlash;
+
+#if ESP_IDF_VERSION_MAJOR >= 5
+        ret = esp_vfs_fat_spiflash_mount_rw_wl(
+            target, source + part_by_name_len, &mount_config, &mount->handle.wl);
+#else
+        ret = esp_vfs_fat_spiflash_mount(
+            target, source + part_by_name_len, &mount_config, &mount->handle.wl);
+#endif
+
+// C3 doesn't support this
+#ifdef SDMMC_SLOT_CONFIG_DEFAULT
+    } else if (!strcmp(source, "sdmmc")) {
+        mount_config.allocation_unit_size = 512;
+
+        sdmmc_host_t host_config = SDMMC_HOST_DEFAULT();
+        sdmmc_slot_config_t slot_config = SDMMC_SLOT_CONFIG_DEFAULT();
+
+        mount = enif_alloc_resource(platform->mounted_fs_resource_type, sizeof(struct MountedFS));
+        if (IS_NULL_PTR(mount)) {
+            RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+        }
+        mount->base_path = strdup(target);
+        mount->mount_type = FATSDMMC;
+
+        ret = esp_vfs_fat_sdmmc_mount(
+            target, &host_config, &slot_config, &mount_config, &mount->handle.card);
+#endif
+
+    } else if (!strcmp(source, "sdspi")) {
+        mount_config.allocation_unit_size = 512;
+
+        sdmmc_host_t host_config = SDSPI_HOST_DEFAULT();
+        sdspi_device_config_t spi_slot_config = SDSPI_DEVICE_CONFIG_DEFAULT();
+
+        term spi_port = interop_kv_get_value_default(
+            opts_term, ATOM_STR("\x8", "spi_host"), term_invalid_term(), ctx->global);
+        spi_host_device_t host_dev;
+        bool ok = spi_driver_get_peripheral(spi_port, &host_dev, ctx->global);
+        if (!ok) {
+            RAISE_ERROR(BADARG_ATOM);
+        }
+        spi_slot_config.host_id = host_dev;
+
+        term cs_term = interop_kv_get_value_default(
+            opts_term, ATOM_STR("\x2", "cs"), term_invalid_term(), ctx->global);
+        VALIDATE_VALUE(cs_term, term_is_integer);
+        spi_slot_config.gpio_cs = term_to_int(cs_term);
+
+        term cd_term = interop_kv_get_value_default(
+            opts_term, ATOM_STR("\x2", "cd"), UNDEFINED_ATOM, ctx->global);
+        if (cd_term != UNDEFINED_ATOM) {
+            VALIDATE_VALUE(cd_term, term_is_integer);
+            spi_slot_config.gpio_cd = term_to_int(cd_term);
+        }
+
+        mount = enif_alloc_resource(platform->mounted_fs_resource_type, sizeof(struct MountedFS));
+        if (IS_NULL_PTR(mount)) {
+            RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+        }
+        mount->base_path = strdup(target);
+        mount->mount_type = FATSDSPI;
+
+        ret = esp_vfs_fat_sdspi_mount(
+            target, &host_config, &spi_slot_config, &mount_config, &mount->handle.card);
```

Comment:
if we move ownership as suggested above, it should not be `target` here but `base_path`.